### ISSUE TITLE
Consider customizable Supervisor/Dispatcher

### ISSUE THREAD

**Issue by kubukoz** [typelevel/cats-effect#1881]
Users have expressed interest in being able to change the behavior of these abstractions when they're being shut down (i.e. waiting for tasks to complete before exiting).

It should be investigated how we can best achieve that.

**Comment by iRevive** [comment]
In my active project, I have a scenario where it's necessary to wait till all fibers are completed before closing the resource. 

Once https://github.com/typelevel/cats-effect/pull/2854 is merged, we will have `parallel` and `sequential` Dispatchers.

Does it make sense to have something similar for `Supervisor`?  For example, `Supervisor.await` and `Supervisor.terminate`?

**Comment by armanbilge** [comment]
I previously suggested https://github.com/typelevel/cats-effect/issues/2537, although maybe a duplicate of this issue :)

**Comment by wjoel** [comment]
Thanks for linking this, I had been incorrectly assuming that `Dispatcher` would wait for tasks to finish, and I think it's part of the confusion mentioned in https://github.com/typelevel/cats-effect/issues/2727 (but there might be more to it).

**Comment by iRevive** [comment]
If I understand use-cases correctly, there are several directions `Dispatcher` is trying to cover:
1) Materialization of an effect. Meanwhile, monad-transformers should be handled in a specific way  https://github.com/typelevel/cats-effect/issues/2448
2) Execution strategy: sequential or parallel
3) Release strategy (basically, depends on the `Supervisor`): terminate pending fibers or backpressure on completion

The 1st point can be accomplished by implicit derivations (as proposed in #2448) and may not require an explicit configuration:
```scala
implicit def eitherTDispatchObserver[F[_]]: DispatcherObserver[EitherT[F, Throwable, *]] = ???
```

For the 2nd and 3rd points a builder can be suitable, I believe.

**Comment by djspiewak** [comment]
Here's what I think we should do. Now that we have `parallel` and `sequential` constructors, I think we should enrich them with some configuration parameters which configure whether lifecycle termination awaits or cancels the inner tasks. Names to be bikeshedded. In practice, this *probably* means that we should also add the same functionality to `Supervisor`, and in fact it's probably the same configuration.

I'd like to get this in for 3.4, since we're already changing the API. This is a very self-contained task, though it's not entirely trivial. I'll mark it as **good first issue** for now, but you'll be touching a few things, so maybe this is more of a "good intermediate issue". :-)

**Comment by iRevive** [comment]
I would like to work on this one. If no one else picked this task up.

**Comment by djspiewak** [comment]
@iRevive Sorry for overlooking your comment here! Please feel free to jump on this!

I've already implemented the `Supervisor` side of it (mostly because I didn't catch your comment), and that's merged in `series/3.x`. I've also implemented an SPSC mode for `Dispatcher` which is in a still-open PR at #2854. The missing piece here is to extend the scope termination functionality from `Supervisor` to be accessible in `Dispatcher` as well, which should be done by building on top of #2854.

I don't rebase branches once I've PR'd them, so it's safe to build on top of that pull request and derive an additional one which adds the configurable scope functionality.

**Comment by iRevive** [comment]
@djspiewak thanks for the suggestion! I will make an update soon
